paymentsheet/src/main/java/com/stripe/android/paymentsheet/forms/Form.kt

```diff
@@ -0,0 +1,149 @@
+                } else if (element is DropdownElement) {
+                    DropDown(element)
+                }
+            }
+        }
+    }
+}
+
+class FormViewModel(
+    val formDataObject: FormDataObject
+) : ViewModel() {
+    class Factory(
+        val formDataObject: FormDataObject
+    ) : ViewModelProvider.Factory {
+        @Suppress("UNCHECKED_CAST")
+        override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+            return FormViewModel(formDataObject) as T
+        }
+    }
+
+    private val paramKey: MutableMap<String, Any?> = formDataObject.paramKey
+    private val types: List<Field> = formDataObject.allTypes
+    private val elementMap = mutableMapOf<Field, Element>()
+    fun getNumberTextFieldElements() = elementMap.filter { it.value is TextFieldElement }.size
+
+    internal fun getElement(type: Field) = requireNotNull(elementMap[type])
+
+    private val isComplete: Flow<Boolean>
+    private val params: Flow<MutableMap<String, Any?>>
+    val paramMapFlow: Flow<MutableMap<String, Any?>?>
+
+    init {
+        val listCompleteFlows = mutableListOf<Flow<Boolean>>()
+        val listOfPairs = mutableListOf<Flow<Pair<String, String?>>>()
+        types.forEach { type ->
+            val element = when (type) {
+                Field.NameInput -> TextFieldElement(NameConfig()) // All configs should have the label passed in for consistency
+                Field.EmailInput -> TextFieldElement(EmailConfig())
+                Field.CountryInput -> DropdownElement(CountryConfig())
+            }
+            listCompleteFlows.add(element.isComplete)
+            listOfPairs.add(element.paymentMethodParams.map {
+                Pair(
+                    type.paymentMethodCreateParamsKey,
+                    it
+                )
+            })
+            elementMap[type] = element
+        }
+
+        isComplete = combine(listCompleteFlows) { elementCompleteState ->
+            elementCompleteState.none { complete -> !complete }
+        }
+
+        params = combine(listOfPairs) { allPairs ->
+            val destMap = mutableMapOf<String, Any?>()
+            createMap(paramKey, destMap, allPairs.toMap())
+            destMap
+        }
+
+        paramMapFlow = combine(isComplete, params) { isComplete, params ->
+            params.takeIf { isComplete }
+        }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private fun createMap(
+        source: Map<String, Any?>,
+        dest: MutableMap<String, Any?>,
+        elementKeys: Map<String, String?>
+    ) {
+        source.keys.forEach { key ->
+            if (source[key] == null) {
+                dest[key] = elementKeys[key]
+            } else if (source[key] is MutableMap<*, *>) {
+                val newDestMap = mutableMapOf<String, Any?>()
+                dest[key] = newDestMap
+                createMap(source[key] as MutableMap<String, Any?>, newDestMap, elementKeys)
+            }
+        }
+    }
```

Comment:
Is this for creating the POST params? `QueryStringFactory` might have the logic you're looking for.

Reply:
Right now I think we need a map to put into the PaymentMethodCreateParams, but it is worth keeping in mind for if we end up down that road later.